src/%IPM/Main.cls

```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	Set allowUpgrade = $$$HasModifier(pCommandInfo,"allow-upgrade")
+	Set useLocal = 1			// var to store the final decision of whether to use local manifest or get from server
+	Set targetVersion = "" 		// var to store the final version of IPM to be installed
+	Kill targetNamespaces 		// multi-dim array to store the final namespaces that need to install IPM
+
+	// basic sanity checks
+	If (quiet && (namespaces = "") && 'globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without specifying namespaces to install."))
+	}
+	If ((namespaces '= "") && globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot specify namespaces and global installation flag at the same time."))
+	}
+
+
+	// 1. Get list of local IPM artifacts under /<instance install dir>/lib/ipm/
+	Set XMLDir = $system.Util.InstallDirectory()
+	For subDirName = "lib", "ipm" {
+		Set XMLDir = ##class(%File).SubDirectoryName(XMLDir, subDirName)
+	}
+  	Set statement = ##class(%SQL.Statement).%New()
+  	Set sc = statement.%PrepareClassQuery("%File", "FileSet")
+  	$$$ThrowOnError(sc)
+	// Valid IPM installation manifest should be of format: ipm-0.0.1.xml
+	Set resultSet = statement.%Execute(XMLDir, "ipm-*.xml")
+	If (resultSet.%SQLCODE < 0)  {
+		$$$ThrowSQLIfError(resultSet.%SQLCODE,resultSet.%Message)
+	}
+	Kill ipmLocalArray			// multi-dim array to store local ipm manifests; ipmLocalArray(<version>)=<fileName>
+	Kill menuList				// menu to let user choose which version of local ipm to install (if not in quiet mode and no version is specified)
+	While resultSet.%Next(.sc) {
+		$$$ThrowOnError(sc)
+		Set fileName = resultSet.%Get("Name")
+		If (resultSet.%Get("Type") = "F") {
+			// get the exact version string from filename
+			Set exactIPMVer = $Piece($Piece(fileName, "ipm-", 2, *), ".xml", 1, *-1)
+			Set ipmLocalArray(exactIPMVer) = fileName
+			Set menuList($Increment(menuList)) = exactIPMVer_" (local version)"
+		}
+	}
+	$$$ThrowOnError(sc)
+
+	If ('$DATA(menuList) && localOnly) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-<version>.xml installer file is found locally in directory: "__XMLDir))
+	}
+
+	// 2. Choose / propmpt user to choose which local/remote version of IPM to install
+	If (version '= "") {
+		Set targetVersion = version
+		If '$DATA(ipmLocalArray(version)) {
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-"_version_".xml installer file is found locally in directory: "_XMLDir))
+			}
+			Write !, "No ipm-"_version_".xml installer file is found in "_XMLDir_" locally. Trying to download from the registry."
+			// need to get from remote registry server
+			Set useLocal = 0
+		}
+	} Else {
+		If quiet {
+			// Non-prompt mode + no version specified - either error out or install latest version from registry
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"To install local IPM in non-prompt mode, a specific version must be specified using the version modifier."))
```

Comment:
Why not just install the highest available version locally? You have all the versions in the multidim array so can use semantic version comparison functions to figure out the highest version.